enhancements/network/user-defined-network-segmentation.md

```diff
@@ -379,11 +382,264 @@ any user-defined primary network.
+
+The CRD status should reflect the user-defined-network state through conditions, 
+reflecting the NAD state and underlying OVN network state.
+
+##### CRD Definitions
+
+The following CRDs should be installed by [CNO](https://github.com/openshift/cluster-network-operator) and managed by OVN-Kubernetes.
+
+##### Namespace scoped CRD
+```golang
+// UserDefinedNetwork describe network request for a Namespace
+type UserDefinedNetwork struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Spec is immutable"
+    // +kubebuilder:validation:XValidation:rule=" self.topology != 'Layer3' || (self.topology == 'Layer3' && size(self.subnets) > 0)", message="Subnets is required for Layer3 topology"
+    // +kubebuilder:validation:XValidation:rule="self.topology != 'Localnet' || (self.topology == 'Localnet' && self.role == 'Secondary')", message="Topology localnet is not supported for primary network"
+    // +kubebuilder:validation:XValidation:rule="!has(self.ipamLifecycle) || (self.ipamLifecycle == 'Persistent' && (self.topology == 'Layer2' || self.topology == 'Localnet'))", message="ipamLifecycle is supported for Layer2 and Localnet topologies"
+    // +required
+    Spec UserDefinedNetworkSpec `json:"spec"`
+    // +optional
+    Status UserDefinedNetworkStatus `json:"status,omitempty"`
+}
+```
+Suggested API validation rules:
+- Spec defined as immutable
+  Avoid incomplete state in a scenario where a NAD is created according to UDN spec, and pods connected to the network, 
+  UDN spec changes, now pods connected to a network that was created from previous revision spec.
+- `Subents` are mandatory for `Layer3` topology.
+- `Localnet` topology is not supported for primary network.
+- `IPAMLifecycle` is supported for `Layer2` and `Localnet` topology.
+
+```golang
+// UserDefinedNetworkSpec defines the desired state of UserDefinedNetwork.
+type UserDefinedNetworkSpec struct {
+	// The topological configuration for the network.
+	// +kubebuilder:validation:Enum=Layer2;Layer3;Localnet
+	Topology NetworkTopology `json:"topology"`
+
+	// The network role in the pod (e.g.: Primary, Secondary).
+	// +kubebuilder:validation:Enum=Primary;Secondary
+	Role NetworkRole `json:"role"`
+
+	// The maximum transmission unit (MTU).
+	// The default value is 1400.
+	// +optional
+	Mtu uint `json:"mtu,omitempty"`
+
+	// The subnet to use for the network across the cluster.
+	// Only include the CIDR for the node. E.g. 10.100.200.0/24.
+	//
+	// IPv6 (2001:DBB::/64) and dual-stack (192.168.100.0/24,2001:DBB::/64) subnets are supported.
+	//
+	// When omitted, the logical switch implementing the network only provides layer 2 communication,
+	// and users must configure IP addresses for the pods.
+	// Port security only prevents MAC spoofing.
+	// +optional
+	Subnets []string `json:"subnets,omitempty"`
```

Comment:
also do we want to limit the number of allowed subnets?

Reply:
The intention was to streamline the CNI subent format by using go native string slice.
Instead of intensifying `"10.0.0.0/16/24"`, the spec should specify `["10.0.0.0/16","10.0.0.0/24"]`
The spec should translate to comma-separated subents list `"10.0.0.0/16,10.0.0.0/24"`.
WDYT?

Reply:
I completely missed where this enhancement says anything about the difference in subnet formats for NAD vs UDn and how it should be converted.
also, won't NAD treat `"10.0.0.0/16,10.0.0.0/24"` as 2 separate networks?

Reply:
is this just a bad example? those 2 separate networks actually overlap and should be invalid.

Reply:
It is not expressed explicitly in the enhancement, sorry for the confusion.
It should be as you said, the CRD `Subents` field value is concated and passed to the CNI in the right format (comma-separated list)

I am not sure what should be the limit for allowed subnets, could you please advise?

Reply:
This is bad example sorry, I tried to express the formatting

Reply:
> should we explain here an L3 format of IP/clusterSubnet/hostSubnet? Also do we want to require hostSubnet for L3 or leave it to default as in NAD?

So getting back to this question ^: do we want to add more comments/examples of correct subnets?
About the length - I think NAD allows any number of subnets, but we may want to limit that in the CRD, e.g. use 2 subnets for dualstack-only, otherwise only 1 (maybe we don't want that, just an example)

Reply:
here is my suggestion:
```
// The subnet to use for the network across the cluster.
// Dualstack clusters may set 2 subnets for each ip family, otherwise only 1 subnet is allowed.
// When topology is `Layer3`, given subnet is split into smaller subnets for every node.
// To specify how the subnet should be split, the following format is supported for `Layer3` network:
// `10.128.0.0/16/24`, which means that every host will get a `/24` subnet.
// If host subnet is not set (for example, `10.128.0.0/16`), it will be set automatically.
// This field may be omitted for `Layer2` and `Localnet` topologies, in that case
// the logical switch implementing the network only provides layer 2 communication,
// and users must configure IP addresses for the pods. Port security only prevents MAC spoofing.
```
used https://github.com/ovn-org/ovn-kubernetes/blob/master/docs/features/multiple-networks/multi-homing.md for reference.

Reply:
I modified it slightly:
```
// The subnet to use for the pod network across the cluster.
// Dualstack clusters may set 2 subnets (one for each ip family), otherwise only 1 subnet is allowed.
// When topology is `Layer3`, the given subnet is split into smaller subnets for every node.
// To specify how the subnet should be split, the following format is supported for `Layer3` network:
// `10.128.0.0/16/24`, which means that every host will get a `/24` subnet.
// If host subnet is not set (for example, `10.128.0.0/16`), it will be assigned automatically.
// This field may be omitted for `Layer2` and `Localnet` topologies, in that case
// the logical switch implementing the network only provides layer 2 communication,
// and users must configure IP addresses for the pods. Port security only prevents MAC spoofing
```

I noticed the above does mention the behavior if the host mask is provided for a layer 2 network. What should be the behavior?

Reply:
Done

Reply:
> I noticed the above does mention the behavior if the host mask is provided for a layer 2 network. What should be the behavior?

It should throw an error. I hoped from the comment it is obvious that host subnet is only supported for Layer3, do you think we can make it more obvious?

Reply:
Maybe just spell out the format for layer 2 and localnet. Like this:

```
// The subnet to use for the pod network across the cluster.
// Dualstack clusters may set 2 subnets (one for each ip family), otherwise only 1 subnet is allowed.
// When topology is `Layer3`, the given subnet is split into smaller subnets for every node.
// To specify how the subnet should be split, the following format is supported for `Layer3` network:
// `10.128.0.0/16/24`, which means that every host will get a `/24` subnet.
// If host subnet mask is not set (for example, `10.128.0.0/16`), it will be assigned automatically.
// For `Layer2` and `Localnet` topology types, the format should match standard CIDR notation, without
// providing any host subnet mask.
// This field may be omitted for `Layer2` and `Localnet` topologies. In that case
// the logical switch implementing the network only provides layer 2 communication,
// and users must configure IP addresses for the pods. Port security only prevents MAC spoofing
```

Reply:
Done